src/metabase/driver/hive_like.clj

```diff
@@ -0,0 +1,160 @@
+      ;; if we need to do an offset we have to do nesting to generate a row number and where on that
+      (let [over-clause (format "row_number() OVER (%s)"
+                                (first (hsql/format (select-keys honeysql-query [:order-by])
+                                                    :allow-dashed-names? true
+                                                    :quoting :mysql)))]
+        (-> (apply h/select (map last (:select honeysql-query)))
+            (h/from (h/merge-select honeysql-query [(hsql/raw over-clause) :__rownum__]))
+            (h/where [:> :__rownum__ offset])
+            (h/limit items))))))
+
+;; lots of copy-paste here. consider making some of those non-private instead.
+(defn- exception->nice-error-message ^String [^java.sql.SQLException e]
+  (or (->> (.getMessage e)     ; error message comes back like 'Column "ZID" not found; SQL statement: ... [error-code]' sometimes
+           (re-find #"^(.*);") ; the user already knows the SQL, and error code is meaningless
+           second)             ; so just return the part of the exception that is relevant
+      (.getMessage e)))
+
+(defn do-with-try-catch {:style/indent 0} [f]
+  (try (f)
+       (catch java.sql.SQLException e
+         (log/error (jdbc/print-sql-exception-chain e))
+         (throw (Exception. (exception->nice-error-message e))))))
+
+(defn- run-query
+  "Run the query itself."
+  [{sql :query, params :params, remark :remark} connection]
+  (let [sql              (str "-- " remark "\n" (hx/unescape-dots sql))
+        statement        (into [sql] params)
+        [columns & rows] (jdbc/query connection statement {:identifiers identity, :as-arrays? true})]
+    {:rows    (or rows [])
+     :columns columns}))
+
+(defn run-query-without-timezone [driver settings connection query]
+  (run-query query connection))
+
+(defprotocol ^:private IUnprepare
+  (^:private unprepare-arg ^String [this]))
+
+(extend-protocol IUnprepare
+  nil     (unprepare-arg [this] "NULL")
+  String  (unprepare-arg [this] (str \' (s/replace this "'" "\\\\'") \')) ; escape single-quotes
+  Boolean (unprepare-arg [this] (if this "TRUE" "FALSE"))
+  Number  (unprepare-arg [this] (str this))
+  Date    (unprepare-arg [this] (first (hsql/format
+                                        (hsql/call :from_unixtime
+                                                   (hsql/call :unix_timestamp
+                                                              (hx/literal (u/date->iso-8601 this))
+                                                              (hx/literal "yyyy-MM-dd\\\\'T\\\\'HH:mm:ss.SSS\\\\'Z\\\\'")))))))
+
+(defn unprepare
+  "Convert a normal SQL `[statement & prepared-statement-args]` vector into a flat, non-prepared statement."
+  ^String [[sql & args]]
+  (loop [sql sql, [arg & more-args, :as args] args]
+    (if-not (seq args)
+      sql
+      (recur (s/replace-first sql #"(?<!\?)\?(?!\?)" (unprepare-arg arg))
+             more-args))))
```

Comment:
same thing as above. Is this needed for a JDBC-based driver?

Reply:
It's slightly different, so it's not just copy-pasta. If there is a way to re-use the common parts of the protocol and modify only parts of it I'd be happy to do so, but I'd need your help to do so, because I don't see it. It has the same name, and the same purpose, but it's tweaked for this driver.

Reply:
@wjoel there's actually a *much* easier way of doing this. You can define custom HoneySQL methods to handle conversion of a function to SQL. So what we can do here is just pass in a custom function to be our `iso-8601-fn` and add a custom handler for that:

```clojure
(require '[honeysql.format :as hformat])

;; custom method to convert a :hive-like-from-unixtime function call to SQL. 
;; hive-like-from-unixtime isn't an actual function, just a name we use so we can pass it to
;; `unprepare` as the `iso-8601-fn` 
(defmethod hformat/fn-handler "hive-like-from-unixtime" [_ datetime-literal]
  (hformat/to-sql
   (hsql/call :from_unixtime
     (hsql/call :unix_timestamp
       datetime-literal
       (hx/literal "yyyy-MM-dd\\\\'T\\\\'HH:mm:ss.SSS\\\\'Z\\\\'")))))

;; (using it)
(unprepare sql-and-args :iso-8601-fn :hive-like-from-unixtime)
```

Let me know if that makes sense. In general any time we're copying entire namespaces to change a single line of code we should look at other ways of doing things to reduce code duplication. We're aiming to keep the driver code as tight and clean as possible because things could quickly spiral out of control in terms of maintenance costs given the number of databases we support

Reply:
Very nice, that seems to work beautifully.